source/common/upstream/health_checker_impl.cc

```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
+  // reset |client_|.
+  //
+  // The right thing to do in to make onInterval() signal to the caller that healthcheck has failed
+  // synchronously, so that timer is not even started.
+  //
+  // Background: before gRPC this was not a problem for HTTP/1.1 healthcheck - it was really hard
+  // to get sync stream reset before connection close. But in gRPC (i.e. HTTP/2) it is very simple -
+  // server signals GOAWAY, preventing well-behaving client from creating new streams. For envoy,
+  // (nghttp2) this practically means sync reset of all Http::StreamEncoder actions. But physically,
+  // HTTP/2 connection is going to be alive for some time.
+  //
+  // Bright side: as long as we drop the connection on healthcheck failures, this case should not
+  // fire.
+  if (client_) {
+    ENVOY_CONN_LOG(debug, "connection/stream timeout health_flags={}", *client_,
+                   HostUtility::healthFlagsToString(*host_));
+  }
+
+  // Someone might guess, how it is possible not to have |request_encoder_| alive (which means there
+  // is no outstanding request) and still hit the timeout. Well, its *possible*. When calling
+  // onInterval(), any of Http::StreamEncoder::encodeHeaders/encodeData may synchronously reset
+  // stream, calling to our onResetStream, which will set |request_encoder_| to null. But as per
+  // above (about |client_| being null), the timeout timer will be started anyway.
+  if (request_encoder_) {
+    // resetState() will be called by onResetStream().
+    expect_reset_ = true;
+    request_encoder_->getStream().resetStream(Http::StreamResetReason::LocalReset);
+  }
+
+  if (client_) {
+    // Force connection close on healthcheck errors and timeouts.
+    // See onResetStream() for rationale.
+    client_->close();
+  }
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::logHealthCheckStatus(
+    Grpc::Status::GrpcStatus grpc_status, const std::string& grpc_message) {
+  const char* service_status;
+  if (!health_check_response_) {
+    service_status = "rpc_error";
+  } else {
+    switch (health_check_response_->status()) {
+    case grpc::health::v1::HealthCheckResponse::SERVING:
+      service_status = "serving";
+      break;
+    case grpc::health::v1::HealthCheckResponse::NOT_SERVING:
+      service_status = "not_serving";
+      break;
+    case grpc::health::v1::HealthCheckResponse::UNKNOWN:
+      service_status = "unknown";
+      break;
+    default:
+      // Should not happen really, Protobuf should not parse undefined enums values.
+      NOT_REACHED;
+      break;
+    }
+  }
+  std::string grpc_status_message;
+  if (grpc_status != Grpc::Status::GrpcStatus::Ok && !grpc_message.empty()) {
+    grpc_status_message = fmt::format("{} ({})", grpc_status, grpc_message);
+  } else {
+    grpc_status_message = fmt::format("{}", grpc_status);
+  }
+  ENVOY_CONN_LOG(debug, "hc grpc_status={} service_status={} health_flags={}", *client_,
```

Comment:
We're doing a bunch of work above that isn't strictly needed when log level is above debug. Is it possible to condition on log level?

Reply:
I haven't found how to do that, but I'm quite new to the codebase. Would appreciate any help :)

Reply:
Just checked with @mrice32, this doesn't exist today unfortunately. I've opened https://github.com/envoyproxy/envoy/issues/2330 to track.